// cojson@0.13.17/dist/coValue.js
export class RawUnknownCoValue {
    constructor(core) {
        this.totalValidTransactions = 0;
        this.id = core.id;
        this.core = core;
    }
    get type() {
        return this.core.header.type;
    }
    get headerMeta() {
        return this.core.header.meta;
    }
    /** @category 6. Meta */
    get group() {
        return this.core.getGroup();
    }
    toJSON() {
        return {};
    }
    atTime() {
        return this;
    }
    subscribe(listener) {
        return this.core.subscribe((content) => {
            listener(content);
        });
    }
}
export function expectMap(content) {
    if (content.type !== "comap") {
        throw new Error("Expected map");
    }
    return content;
}
export function expectList(content) {
    if (content.type !== "colist") {
        throw new Error("Expected list");
    }
    return content;
}
export function expectStream(content) {
    if (content.type !== "costream") {
        throw new Error("Expected stream");
    }
    return content;
}
export function expectPlainText(content) {
    if (content.type !== "coplaintext") {
        throw new Error("Expected plaintext");
    }
    return content;
}
